client/beefy/src/worker.rs

```diff
@@ -65,6 +69,34 @@ where
 	pub metrics: Option<Metrics>,
 }
 
+#[derive(Default)]
+struct BeefyTicker {
+	inner: Arc<Mutex<(bool, Option<Waker>)>>,
+}
+
+impl BeefyTicker {
+	pub fn wake(&mut self) {
+		let mut guard = self.inner.lock();
+		guard.0 = true;
+		guard.1.take().map(|waker| waker.wake());
+	}
+}
```

Comment:
I agree that `#3` is weird and shouldn't be considered. `#2` as it is implemented right now is also "inefficient" in the sense that you wake the voter but you don't know the reason it was awakened, e.g. whenever we get a new vote we wake the voter which starts by checking if there's a new session (which we know didn't happen, we just got a vote). But more than efficiency concerns I think this makes the code less readable. 

My main point was that with option `#1` we don't need to have such a coarse `tick()` function where we do everything, e.g. if there's a finalized block and it has the authorities change digest then we can have the worker react to this and only do whatever it needs to do. 

We can achieve the same thing with option `#2` (and I am OK with it) if instead of having coarse wake up logic we have internal events that we inject into some channel, then the worker can listen to these events (and automatically wake up based on them without explicit waker) and it will know the reason why it was awakened (e.g. there's a new vote, authorities have changed, we got a justification/signed commitment) and thus react in a smarter way. We can also pass some data along in these events (that might be required for the worker to do its job) which means that there will be less data sharing between different components.

Reply:
yes, that sounds great! I'll rework it in that direction.